# **Chapter 4 & 5 - Subqueries & Table Expressions - Homework 3**

## **Group 3** - _**Damian Rozpedowski**, Hannah Kurian, Hasnatul Hosna, Essmer Sanchez, Zicheng Yang_

### _Due 2/26/2024_

## **\========== Chapter 4 Subqueries ==========**

**Proposition**: Identify the details of the latest order placed.  
**Tables**: Sales.\[Order\]  
**Columns**: orderid, orderdate, employeeid, customerid  
**Predicate**: The latest order is determined by the highest order ID.

In [22]:
---------------------------------------------------------------------
-- Scalar Subqueries
---------------------------------------------------------------------

-- Order with the maximum order ID
USE TSQLV6;
DECLARE @maxid AS INT = (SELECT MAX(orderid)
                         FROM Sales.Orders);

SELECT orderid, orderdate, empid, custid
FROM Sales.Orders
WHERE orderid = @maxid;

USE Northwinds2022TSQLV7
DECLARE @maxid2 AS INT = (SELECT MAX(orderid)
                         FROM Sales.[Order]);
SELECT orderid, orderdate, employeeid, customerid
FROM Sales.[Order]
WHERE orderid = @maxid2;

(1 row affected)

(1 row affected)

Total execution time: 00:00:00.011

orderid orderdate empid custid 11077 2022-05-06 1 65

orderid orderdate employeeid customerid 11077 2016-05-06 1 65

**Proposition**<span style="background-color: rgb(255, 255, 255); color: rgb(13, 13, 13); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">: Retrieve the most recent order's details.<br></span>**Tables**: Sales.\[Order\]  
**Columns**: orderid, orderdate, employeeid, customerid  
**Predicate**: Selects the order with the maximum order ID.

In [23]:
USE TSQLV6
SELECT orderid, orderdate, empid, custid
FROM Sales.Orders
WHERE orderid = (SELECT MAX(O.orderid)
                 FROM Sales.Orders AS O);

USE Northwinds2022TSQLV7
SELECT orderid, orderdate, employeeid, customerid
FROM Sales.[Order]
WHERE orderid = (SELECT MAX(O.orderid)
                 FROM Sales.[Order] AS O);

(1 row affected)

(1 row affected)

Total execution time: 00:00:00.008

orderid orderdate empid custid 11077 2022-05-06 1 65

orderid orderdate employeeid customerid 11077 2016-05-06 1 65

**Proposition**: Find all orders handled by employees whose last names start with "C".

**Tables**: Sales.\[Order\], HumanResources.Employee

**Columns**: orderid, employeeid, employeelastname

**Predicate**: Matches orders to employees with last names beginning with "C".

In [24]:
USE TSQLV6
-- Scalar subquery expected to return one value
SELECT orderid
FROM Sales.Orders
WHERE empid = 
  (SELECT E.empid
   FROM HR.Employees AS E
   WHERE E.lastname LIKE N'C%');

USE Northwinds2022TSQLV7
SELECT orderid
FROM Sales.[Order]
WHERE employeeid = 
  (SELECT E.employeeid
   FROM HumanResources.Employee AS E
   WHERE E.employeelastname LIKE N'C%');

(104 rows affected)

(104 rows affected)

Total execution time: 00:00:00.106

orderid 10262 10268 10276 10278 10279 10286 10287 10290 10301 10305 10310 10318 10334 10354 10366 10369 10380 10383 10399 10402 10408 10412 10416 10421 10435 10437 10443 10450 10452 10455 10456 10460 10467 10472 10476 10481 10488 10491 10498 10521 10533 10534 10543 10545 10560 10565 10571 10589 10596 10602 10603 10610 10614 10623 10627 10631 10632 10635 10651 10660 10679 10694 10696 10706 10719 10720 10722 10724 10729 10756 10770 10786 10795 10811 10824 10844 10845 10852 10857 10862 10883 10887 10932 10940 10955 10957 10961 10962 10977 10979 10986 10987 10997 10998 11007 11034 11036 11046 11050 11054 11056 11065 11068 11075

orderid 10262 10268 10276 10278 10279 10286 10287 10290 10301 10305 10310 10318 10334 10354 10366 10369 10380 10383 10399 10402 10408 10412 10416 10421 10435 10437 10443 10450 10452 10455 10456 10460 10467 10472 10476 10481 10488 10491 10498 10521 10533 10534 10543 10545 10560 10565 10571 10589 10596 10602 10603 10610 10614 10623 10627 10631 10632 10635 10651 10660 10679 10694 10696 10706 10719 10720 10722 10724 10729 10756 10770 10786 10795 10811 10824 10844 10845 10852 10857 10862 10883 10887 10932 10940 10955 10957 10961 10962 10977 10979 10986 10987 10997 10998 11007 11034 11036 11046 11050 11054 11056 11065 11068 11075

**Proposition**: Identify orders processed by employees with last names starting with "D".

**Tables**: Sales.\[Order\], HumanResources.Employee

**Columns**: orderid, employeeid, employeelastname

**Predicate**: The employee's last name begins with "D".

In [25]:
USE TSQLV6
SELECT orderid
FROM Sales.Orders
WHERE empid = 
  (SELECT E.empid
   FROM HR.Employees AS E
   WHERE E.lastname LIKE N'D%');

USE Northwinds2022TSQLV7
SELECT orderid
FROM Sales.[Order]
WHERE employeeid = 
  (SELECT E.employeeid
   FROM HumanResources.Employee AS E
   WHERE E.employeelastname LIKE N'D%');

Total execution time: 00:00:00.008

orderid

orderid

: Msg 512, Level 16, State 1, Line 2
Subquery returned more than 1 value. This is not permitted when the subquery follows =, !=, <, <= , >, >= or when the subquery is used as an expression.

: Msg 512, Level 16, State 1, Line 10
Subquery returned more than 1 value. This is not permitted when the subquery follows =, !=, <, <= , >, >= or when the subquery is used as an expression.

**Proposition:**   <span style="color: var(--vscode-foreground);">&nbsp;Find orders managed by employees whose last names start with "A".</span>

**Tables**<span style="color: var(--vscode-foreground);">: Sales.[Order], HumanResources.Employee</span>

**Columns**: orderid (from Sales.\[Order\]), employeeid, employeelastname (from HumanResources.Employee)

**Predicate**<span style="color: var(--vscode-foreground);">: Employees' last names start with "A".</span>

In [26]:
USE TSQLV6
SELECT orderid
FROM Sales.Orders
WHERE empid = 
  (SELECT E.empid
   FROM HR.Employees AS E
   WHERE E.lastname LIKE N'A%');

USE Northwinds2022TSQLV7
SELECT orderid
FROM Sales.[Order]
WHERE employeeid = 
  (SELECT E.employeeid
   FROM HumanResources.Employee AS E
   WHERE E.employeelastname LIKE N'A%');

(0 rows affected)

(0 rows affected)

Total execution time: 00:00:00.008

orderid

orderid

## **\========== Chapter 4 Subqueries - Excercises ==========**

\-- 4

\-- Write a query that returns

\-- countries where there are customers but not employees

\-- Tables involved: TSQLV6 database, Customers and Employees tables

\-- Desired output:

country

\---------------

Argentina

Austria

Belgium

Brazil

Canada

Denmark

Finland

France

Germany

Ireland

Italy

Mexico

Norway

Poland

Portugal

Spain

Sweden

Switzerland

Venezuela

(19 rows affected)

<span style="color: var(--tw-prose-bold); border: 0px solid rgb(227, 227, 227); box-sizing: border-box; --tw-border-spacing-x: 0; --tw-border-spacing-y: 0; --tw-translate-x: 0; --tw-translate-y: 0; --tw-rotate: 0; --tw-skew-x: 0; --tw-skew-y: 0; --tw-scale-x: 1; --tw-scale-y: 1; --tw-pan-x: ; --tw-pan-y: ; --tw-pinch-zoom: ; --tw-scroll-snap-strictness: proximity; --tw-gradient-from-position: ; --tw-gradient-via-position: ; --tw-gradient-to-position: ; --tw-ordinal: ; --tw-slashed-zero: ; --tw-numeric-figure: ; --tw-numeric-spacing: ; --tw-numeric-fraction: ; --tw-ring-inset: ; --tw-ring-offset-width: 0px; --tw-ring-offset-color: #fff; --tw-ring-color: rgba(69,89,164,.5); --tw-ring-offset-shadow: 0 0 transparent; --tw-ring-shadow: 0 0 transparent; --tw-shadow: 0 0 transparent; --tw-shadow-colored: 0 0 transparent; --tw-blur: ; --tw-brightness: ; --tw-contrast: ; --tw-grayscale: ; --tw-hue-rotate: ; --tw-invert: ; --tw-saturate: ; --tw-sepia: ; --tw-drop-shadow: ; --tw-backdrop-blur: ; --tw-backdrop-brightness: ; --tw-backdrop-contrast: ; --tw-backdrop-grayscale: ; --tw-backdrop-hue-rotate: ; --tw-backdrop-invert: ; --tw-backdrop-opacity: ; --tw-backdrop-saturate: ; --tw-backdrop-sepia: ; font-weight: 600;"><br></span>

<span style="color: var(--tw-prose-bold); border: 0px solid rgb(227, 227, 227); box-sizing: border-box; --tw-border-spacing-x: 0; --tw-border-spacing-y: 0; --tw-translate-x: 0; --tw-translate-y: 0; --tw-rotate: 0; --tw-skew-x: 0; --tw-skew-y: 0; --tw-scale-x: 1; --tw-scale-y: 1; --tw-pan-x: ; --tw-pan-y: ; --tw-pinch-zoom: ; --tw-scroll-snap-strictness: proximity; --tw-gradient-from-position: ; --tw-gradient-via-position: ; --tw-gradient-to-position: ; --tw-ordinal: ; --tw-slashed-zero: ; --tw-numeric-figure: ; --tw-numeric-spacing: ; --tw-numeric-fraction: ; --tw-ring-inset: ; --tw-ring-offset-width: 0px; --tw-ring-offset-color: #fff; --tw-ring-color: rgba(69,89,164,.5); --tw-ring-offset-shadow: 0 0 transparent; --tw-ring-shadow: 0 0 transparent; --tw-shadow: 0 0 transparent; --tw-shadow-colored: 0 0 transparent; --tw-blur: ; --tw-brightness: ; --tw-contrast: ; --tw-grayscale: ; --tw-hue-rotate: ; --tw-invert: ; --tw-saturate: ; --tw-sepia: ; --tw-drop-shadow: ; --tw-backdrop-blur: ; --tw-backdrop-brightness: ; --tw-backdrop-contrast: ; --tw-backdrop-grayscale: ; --tw-backdrop-hue-rotate: ; --tw-backdrop-invert: ; --tw-backdrop-opacity: ; --tw-backdrop-saturate: ; --tw-backdrop-sepia: ;"><b>Proposition</b></span><span style="color: var(--tw-prose-bold); border: 0px solid rgb(227, 227, 227); box-sizing: border-box; --tw-border-spacing-x: 0; --tw-border-spacing-y: 0; --tw-translate-x: 0; --tw-translate-y: 0; --tw-rotate: 0; --tw-skew-x: 0; --tw-skew-y: 0; --tw-scale-x: 1; --tw-scale-y: 1; --tw-pan-x: ; --tw-pan-y: ; --tw-pinch-zoom: ; --tw-scroll-snap-strictness: proximity; --tw-gradient-from-position: ; --tw-gradient-via-position: ; --tw-gradient-to-position: ; --tw-ordinal: ; --tw-slashed-zero: ; --tw-numeric-figure: ; --tw-numeric-spacing: ; --tw-numeric-fraction: ; --tw-ring-inset: ; --tw-ring-offset-width: 0px; --tw-ring-offset-color: #fff; --tw-ring-color: rgba(69,89,164,.5); --tw-ring-offset-shadow: 0 0 transparent; --tw-ring-shadow: 0 0 transparent; --tw-shadow: 0 0 transparent; --tw-shadow-colored: 0 0 transparent; --tw-blur: ; --tw-brightness: ; --tw-contrast: ; --tw-grayscale: ; --tw-hue-rotate: ; --tw-invert: ; --tw-saturate: ; --tw-sepia: ; --tw-drop-shadow: ; --tw-backdrop-blur: ; --tw-backdrop-brightness: ; --tw-backdrop-contrast: ; --tw-backdrop-grayscale: ; --tw-backdrop-hue-rotate: ; --tw-backdrop-invert: ; --tw-backdrop-opacity: ; --tw-backdrop-saturate: ; --tw-backdrop-sepia: ; font-weight: 600;">:&nbsp;</span>  <span style="color: var(--vscode-foreground);">Write a query that returns&nbsp;</span>  <span style="color: var(--vscode-foreground);">countries where there are customers but not employees</span>

**Tables**: Sales.Customer, HumanResources.Employee

**Columns**: customerCountry (from Sales.Customer), employeeCountry (from HumanResources.Employee)

**Predicate**: Exclude countries where customers and employees share the same country.

In [27]:
USE TSQLV6
SELECT DISTINCT c.country
FROM Sales.Customers c
WHERE c.country NOT IN (
    SELECT DISTINCT e.Country
    FROM HR.Employees e
)
ORDER BY c.country;

USE Northwinds2022TSQLV7
SELECT DISTINCT c.customerCountry
FROM Sales.Customer c
WHERE c.customerCountry NOT IN (
    SELECT DISTINCT e.employeeCountry
    FROM HumanResources.Employee e
)
ORDER BY c.customerCountry;

(19 rows affected)

(19 rows affected)

Total execution time: 00:00:00.110

country Argentina Austria Belgium Brazil Canada Denmark Finland France Germany Ireland Italy Mexico Norway Poland Portugal Spain Sweden Switzerland Venezuela

customerCountry Argentina Austria Belgium Brazil Canada Denmark Finland France Germany Ireland Italy Mexico Norway Poland Portugal Spain Sweden Switzerland Venezuela

## **\========== Chapter 5 Table Expressions ==========**

**Proposition**: Display all customer IDs and company names for customers located in the USA.

**Tables**: Sales.Customer

**Columns**: customerid, customerCompanyName

**Predicate**: The customer is located in the USA.

In [28]:
---------------------------------------------------------------------
-- Derived Tables
---------------------------------------------------------------------
USE TSQLV6;
SELECT *
FROM (SELECT custid, companyname
      FROM Sales.Customers
      WHERE country = N'USA') AS USACusts;

USE Northwinds2022TSQLV7
SELECT *
FROM (SELECT customerid, customerCompanyName
      FROM Sales.Customer
      WHERE customerCountry = N'USA') AS USACusts;

(13 rows affected)

(13 rows affected)

Total execution time: 00:00:00.009

custid companyname 32 Customer YSIQX 36 Customer LVJSO 43 Customer UISOJ 45 Customer QXPPT 48 Customer DVFMB 55 Customer KZQZT 65 Customer NYUHS 71 Customer LCOUJ 75 Customer XOJYP 77 Customer LCYBZ 78 Customer NLTYP 82 Customer EYHKM 89 Customer YBQTI

customerid customerCompanyName 32 Customer YSIQX 36 Customer LVJSO 43 Customer UISOJ 45 Customer QXPPT 48 Customer DVFMB 55 Customer KZQZT 65 Customer NYUHS 71 Customer LCOUJ 75 Customer XOJYP 77 Customer LCYBZ 78 Customer NLTYP 82 Customer EYHKM 89 Customer YBQTI

**Proposition**: Count the number of unique customers placing orders each year.

**Tables**: Sales.\[Order\]

**Columns**: orderdate, customerid, numcusts

**Predicate**: No predicate.

In [29]:
USE TSQLV6;
-- Listing 5-1 Query with a Derived Table using Inline Aliasing Form
SELECT orderyear, COUNT(DISTINCT custid) AS numcusts
FROM (SELECT YEAR(orderdate) AS orderyear, custid
      FROM Sales.Orders) AS D
GROUP BY orderyear;

USE Northwinds2022TSQLV7
SELECT orderyear, COUNT(DISTINCT customerid) AS numcusts
FROM (SELECT YEAR(orderdate) AS orderyear, customerid
      FROM Sales.[Order]) AS D
GROUP BY orderyear;

(3 rows affected)

(3 rows affected)

Total execution time: 00:00:00.018

orderyear numcusts 2020 67 2021 86 2022 81

orderyear numcusts 2014 67 2015 86 2016 81

<span style="color: var(--tw-prose-bold); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve; border: 0px solid rgb(227, 227, 227); box-sizing: border-box; --tw-border-spacing-x: 0; --tw-border-spacing-y: 0; --tw-translate-x: 0; --tw-translate-y: 0; --tw-rotate: 0; --tw-skew-x: 0; --tw-skew-y: 0; --tw-scale-x: 1; --tw-scale-y: 1; --tw-pan-x: ; --tw-pan-y: ; --tw-pinch-zoom: ; --tw-scroll-snap-strictness: proximity; --tw-gradient-from-position: ; --tw-gradient-via-position: ; --tw-gradient-to-position: ; --tw-ordinal: ; --tw-slashed-zero: ; --tw-numeric-figure: ; --tw-numeric-spacing: ; --tw-numeric-fraction: ; --tw-ring-inset: ; --tw-ring-offset-width: 0px; --tw-ring-offset-color: #fff; --tw-ring-color: rgba(69,89,164,.5); --tw-ring-offset-shadow: 0 0 transparent; --tw-ring-shadow: 0 0 transparent; --tw-shadow: 0 0 transparent; --tw-shadow-colored: 0 0 transparent; --tw-blur: ; --tw-brightness: ; --tw-contrast: ; --tw-grayscale: ; --tw-hue-rotate: ; --tw-invert: ; --tw-saturate: ; --tw-sepia: ; --tw-drop-shadow: ; --tw-backdrop-blur: ; --tw-backdrop-brightness: ; --tw-backdrop-contrast: ; --tw-backdrop-grayscale: ; --tw-backdrop-hue-rotate: ; --tw-backdrop-invert: ; --tw-backdrop-opacity: ; --tw-backdrop-saturate: ; --tw-backdrop-sepia: ; font-weight: 600;">Proposition</span><span style="background-color: rgb(255, 255, 255); color: rgb(13, 13, 13); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">: Count unique customers per year based on their order dates.</span>  

<span style="border: 0px solid rgb(227, 227, 227); box-sizing: border-box; --tw-border-spacing-x: 0; --tw-border-spacing-y: 0; --tw-translate-x: 0; --tw-translate-y: 0; --tw-rotate: 0; --tw-skew-x: 0; --tw-skew-y: 0; --tw-scale-x: 1; --tw-scale-y: 1; --tw-pan-x: ; --tw-pan-y: ; --tw-pinch-zoom: ; --tw-scroll-snap-strictness: proximity; --tw-gradient-from-position: ; --tw-gradient-via-position: ; --tw-gradient-to-position: ; --tw-ordinal: ; --tw-slashed-zero: ; --tw-numeric-figure: ; --tw-numeric-spacing: ; --tw-numeric-fraction: ; --tw-ring-inset: ; --tw-ring-offset-width: 0px; --tw-ring-offset-color: #fff; --tw-ring-color: rgba(69,89,164,.5); --tw-ring-offset-shadow: 0 0 transparent; --tw-ring-shadow: 0 0 transparent; --tw-shadow: 0 0 transparent; --tw-shadow-colored: 0 0 transparent; --tw-blur: ; --tw-brightness: ; --tw-contrast: ; --tw-grayscale: ; --tw-hue-rotate: ; --tw-invert: ; --tw-saturate: ; --tw-sepia: ; --tw-drop-shadow: ; --tw-backdrop-blur: ; --tw-backdrop-brightness: ; --tw-backdrop-contrast: ; --tw-backdrop-grayscale: ; --tw-backdrop-hue-rotate: ; --tw-backdrop-invert: ; --tw-backdrop-opacity: ; --tw-backdrop-saturate: ; --tw-backdrop-sepia: ; font-weight: 600; color: var(--tw-prose-bold);">Tables</span>: Sales.\[Order\]

<span style="border: 0px solid rgb(227, 227, 227); box-sizing: border-box; --tw-border-spacing-x: 0; --tw-border-spacing-y: 0; --tw-translate-x: 0; --tw-translate-y: 0; --tw-rotate: 0; --tw-skew-x: 0; --tw-skew-y: 0; --tw-scale-x: 1; --tw-scale-y: 1; --tw-pan-x: ; --tw-pan-y: ; --tw-pinch-zoom: ; --tw-scroll-snap-strictness: proximity; --tw-gradient-from-position: ; --tw-gradient-via-position: ; --tw-gradient-to-position: ; --tw-ordinal: ; --tw-slashed-zero: ; --tw-numeric-figure: ; --tw-numeric-spacing: ; --tw-numeric-fraction: ; --tw-ring-inset: ; --tw-ring-offset-width: 0px; --tw-ring-offset-color: #fff; --tw-ring-color: rgba(69,89,164,.5); --tw-ring-offset-shadow: 0 0 transparent; --tw-ring-shadow: 0 0 transparent; --tw-shadow: 0 0 transparent; --tw-shadow-colored: 0 0 transparent; --tw-blur: ; --tw-brightness: ; --tw-contrast: ; --tw-grayscale: ; --tw-hue-rotate: ; --tw-invert: ; --tw-saturate: ; --tw-sepia: ; --tw-drop-shadow: ; --tw-backdrop-blur: ; --tw-backdrop-brightness: ; --tw-backdrop-contrast: ; --tw-backdrop-grayscale: ; --tw-backdrop-hue-rotate: ; --tw-backdrop-invert: ; --tw-backdrop-opacity: ; --tw-backdrop-saturate: ; --tw-backdrop-sepia: ; font-weight: 600; color: var(--tw-prose-bold);">Columns</span>: orderyear, orderdate, customerid, numcusts

<span style="border: 0px solid rgb(227, 227, 227); box-sizing: border-box; --tw-border-spacing-x: 0; --tw-border-spacing-y: 0; --tw-translate-x: 0; --tw-translate-y: 0; --tw-rotate: 0; --tw-skew-x: 0; --tw-skew-y: 0; --tw-scale-x: 1; --tw-scale-y: 1; --tw-pan-x: ; --tw-pan-y: ; --tw-pinch-zoom: ; --tw-scroll-snap-strictness: proximity; --tw-gradient-from-position: ; --tw-gradient-via-position: ; --tw-gradient-to-position: ; --tw-ordinal: ; --tw-slashed-zero: ; --tw-numeric-figure: ; --tw-numeric-spacing: ; --tw-numeric-fraction: ; --tw-ring-inset: ; --tw-ring-offset-width: 0px; --tw-ring-offset-color: #fff; --tw-ring-color: rgba(69,89,164,.5); --tw-ring-offset-shadow: 0 0 transparent; --tw-ring-shadow: 0 0 transparent; --tw-shadow: 0 0 transparent; --tw-shadow-colored: 0 0 transparent; --tw-blur: ; --tw-brightness: ; --tw-contrast: ; --tw-grayscale: ; --tw-hue-rotate: ; --tw-invert: ; --tw-saturate: ; --tw-sepia: ; --tw-drop-shadow: ; --tw-backdrop-blur: ; --tw-backdrop-brightness: ; --tw-backdrop-contrast: ; --tw-backdrop-grayscale: ; --tw-backdrop-hue-rotate: ; --tw-backdrop-invert: ; --tw-backdrop-opacity: ; --tw-backdrop-saturate: ; --tw-backdrop-sepia: ; font-weight: 600; color: var(--tw-prose-bold);">Predicate</span>: No predicate.

In [30]:
USE TSQLV6;
SELECT YEAR(orderdate) AS orderyear, COUNT(DISTINCT custid) AS numcusts
FROM Sales.Orders
GROUP BY YEAR(orderdate);

USE Northwinds2022TSQLV7
SELECT YEAR(orderdate) AS orderyear, COUNT(DISTINCT customerid) AS numcusts
FROM Sales.[Order]
GROUP BY YEAR(orderdate);

(3 rows affected)

(3 rows affected)

Total execution time: 00:00:00.065

orderyear numcusts 2020 67 2021 86 2022 81

orderyear numcusts 2014 67 2015 86 2016 81

**Proposition**: Determine the annual count of unique customers based on order dates.

**Tables**: Sales.\[Order\]

**Columns**: orderyear, customerid

**Predicate**: No predicate.

In [31]:
USE TSQLV6;
-- External column aliasing
SELECT orderyear, COUNT(DISTINCT custid) AS numcusts
FROM (SELECT YEAR(orderdate), custid
      FROM Sales.Orders) AS D(orderyear, custid)
GROUP BY orderyear;
GO

USE Northwinds2022TSQLV7
SELECT orderyear, COUNT(DISTINCT custid) AS numcusts
FROM (SELECT YEAR(orderdate), customerid
      FROM Sales.[Order]) AS D(orderyear, custid)
GROUP BY orderyear;
GO

(3 rows affected)

(3 rows affected)

Total execution time: 00:00:00.015

orderyear numcusts 2020 67 2021 86 2022 81

orderyear numcusts 2014 67 2015 86 2016 81

**Proposition**: Calculate the number of unique customers per year for orders handled by a specific employee.

**Tables**: Sales.\[Order\]

**Columns**: orderyear, customerid

**Predicate**: Filters orders by a specific employee ID.

In [32]:
USE TSQLV6;
---------------------------------------------------------------------
-- Using Arguments
---------------------------------------------------------------------

-- Yearly Count of Customers handled by Employee 3
DECLARE @empid AS INT = 3;

SELECT orderyear, COUNT(DISTINCT custid) AS numcusts
FROM (SELECT YEAR(orderdate) AS orderyear, custid
      FROM Sales.Orders
      WHERE empid = @empid) AS D
GROUP BY orderyear;
GO

USE Northwinds2022TSQLV7
DECLARE @employeeid AS INT = 3;

SELECT orderyear, COUNT(DISTINCT customerid) AS numcusts
FROM (SELECT YEAR(orderdate) AS orderyear, customerid
      FROM Sales.[Order]
      WHERE employeeid = @employeeid) AS D
GROUP BY orderyear;
GO

(3 rows affected)

(3 rows affected)

Total execution time: 00:00:00.017

orderyear numcusts 2020 16 2021 46 2022 30

orderyear numcusts 2014 16 2015 46 2016 30

**Proposition**: Identify years with more than 70 unique customers placing orders.

**Tables**: Sales.\[Order\]

**Columns**: orderyear, customerid

**Predicate**: Years having more than 70 distinct customers.

In [33]:
USE TSQLV6;
-- Listing 5-2 Query with Nested Derived Tables
SELECT orderyear, numcusts
FROM (SELECT orderyear, COUNT(DISTINCT custid) AS numcusts
      FROM (SELECT YEAR(orderdate) AS orderyear, custid
            FROM Sales.Orders) AS D1
      GROUP BY orderyear) AS D2
WHERE numcusts > 70;

USE Northwinds2022TSQLV7
SELECT orderyear, numcusts
FROM (SELECT orderyear, COUNT(DISTINCT customerid) AS numcusts
      FROM (SELECT YEAR(orderdate) AS orderyear, customerid
            FROM Sales.[Order]) AS D1
      GROUP BY orderyear) AS D2
WHERE numcusts > 70;

(2 rows affected)

(2 rows affected)

Total execution time: 00:00:00.014

orderyear numcusts 2021 86 2022 81

orderyear numcusts 2015 86 2016 81

**Proposition**: Find years where over 70 unique customers placed orders.

**Tables**: Sales.\[Order\]

**Columns**: orderyear, customerid

**Predicate**: Only includes years with more than 70 distinct customers.

In [34]:
USE TSQLV6;
SELECT YEAR(orderdate) AS orderyear, COUNT(DISTINCT custid) AS numcusts
FROM Sales.Orders
GROUP BY YEAR(orderdate)
HAVING COUNT(DISTINCT custid) > 70;

USE Northwinds2022TSQLV7
SELECT YEAR(orderdate) AS orderyear, COUNT(DISTINCT customerid) AS numcusts
FROM Sales.[Order]
GROUP BY YEAR(orderdate)
HAVING COUNT(DISTINCT customerid) > 70;

(2 rows affected)

(2 rows affected)

Total execution time: 00:00:00.014

orderyear numcusts 2021 86 2022 81

orderyear numcusts 2015 86 2016 81

**Proposition**<span style="background-color: rgb(255, 255, 255); color: rgb(13, 13, 13); font-family: Söhne, ui-sans-serif, system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, &quot;Helvetica Neue&quot;, Arial, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 16px; white-space-collapse: preserve;">: Compare the annual growth in the number of unique customers placing orders, year over year.</span>

**Tables**<span style="color: var(--vscode-foreground);">: Sales.[Order]</span>

**Columns**: orderyear, customerid

**Predicate**<span style="color: var(--vscode-foreground);">: Matches each year with the previous year to calculate growth in unique customers.</span>

In [35]:
USE TSQLV6;
---------------------------------------------------------------------
-- Multiple References
---------------------------------------------------------------------

-- Listing 5-3 Multiple Derived Tables Based on the Same Query
SELECT Cur.orderyear, 
  Cur.numcusts AS curnumcusts, Prv.numcusts AS prvnumcusts,
  Cur.numcusts - Prv.numcusts AS growth
FROM (SELECT YEAR(orderdate) AS orderyear,
        COUNT(DISTINCT custid) AS numcusts
      FROM Sales.Orders
      GROUP BY YEAR(orderdate)) AS Cur
  LEFT OUTER JOIN
     (SELECT YEAR(orderdate) AS orderyear,
        COUNT(DISTINCT custid) AS numcusts
      FROM Sales.Orders
      GROUP BY YEAR(orderdate)) AS Prv
    ON Cur.orderyear = Prv.orderyear + 1;

USE Northwinds2022TSQLV7
SELECT Cur.orderyear, 
  Cur.numcusts AS curnumcusts, Prv.numcusts AS prvnumcusts,
  Cur.numcusts - Prv.numcusts AS growth
FROM (SELECT YEAR(orderdate) AS orderyear,
        COUNT(DISTINCT customerid) AS numcusts
      FROM Sales.[Order]
      GROUP BY YEAR(orderdate)) AS Cur
  LEFT OUTER JOIN
     (SELECT YEAR(orderdate) AS orderyear,
        COUNT(DISTINCT customerid) AS numcusts
      FROM Sales.[Order]
      GROUP BY YEAR(orderdate)) AS Prv
    ON Cur.orderyear = Prv.orderyear + 1;

(3 rows affected)

(3 rows affected)

Total execution time: 00:00:00.096

orderyear curnumcusts prvnumcusts growth 2020 67 NULL NULL 2021 86 67 19 2022 81 86 -5

orderyear curnumcusts prvnumcusts growth 2014 67 NULL NULL 2015 86 67 19 2016 81 86 -5

## **\========== Chapter 5 Table Expressions - Excercises ==========**

\-- 3-1

\-- Write a query that calculates a row number for each order

\-- based on orderdate, orderid ordering

\-- Tables involved: Sales.Orders

\-- Desired output:

orderid     orderdate   custid      empid       rownum

\----------- ----------- ----------- ----------- -------

10248       2020-07-04  85          5           1

10249       2020-07-05  79          6           2

10250       2020-07-08  34          4           3

10251       2020-07-08  84          3           4

10252       2020-07-09  76          4           5

10253       2020-07-10  34          3           6

10254       2020-07-11  14          5           7

10255       2020-07-12  68          9           8

10256       2020-07-15  88          3           9

10257       2020-07-16  35          4           10

...

(830 rows affected)

<span style="color: var(--tw-prose-bold); border: 0px solid rgb(227, 227, 227); box-sizing: border-box; --tw-border-spacing-x: 0; --tw-border-spacing-y: 0; --tw-translate-x: 0; --tw-translate-y: 0; --tw-rotate: 0; --tw-skew-x: 0; --tw-skew-y: 0; --tw-scale-x: 1; --tw-scale-y: 1; --tw-pan-x: ; --tw-pan-y: ; --tw-pinch-zoom: ; --tw-scroll-snap-strictness: proximity; --tw-gradient-from-position: ; --tw-gradient-via-position: ; --tw-gradient-to-position: ; --tw-ordinal: ; --tw-slashed-zero: ; --tw-numeric-figure: ; --tw-numeric-spacing: ; --tw-numeric-fraction: ; --tw-ring-inset: ; --tw-ring-offset-width: 0px; --tw-ring-offset-color: #fff; --tw-ring-color: rgba(69,89,164,.5); --tw-ring-offset-shadow: 0 0 transparent; --tw-ring-shadow: 0 0 transparent; --tw-shadow: 0 0 transparent; --tw-shadow-colored: 0 0 transparent; --tw-blur: ; --tw-brightness: ; --tw-contrast: ; --tw-grayscale: ; --tw-hue-rotate: ; --tw-invert: ; --tw-saturate: ; --tw-sepia: ; --tw-drop-shadow: ; --tw-backdrop-blur: ; --tw-backdrop-brightness: ; --tw-backdrop-contrast: ; --tw-backdrop-grayscale: ; --tw-backdrop-hue-rotate: ; --tw-backdrop-invert: ; --tw-backdrop-opacity: ; --tw-backdrop-saturate: ; --tw-backdrop-sepia: ; font-weight: 600;">Proposition:&nbsp;</span> <span style="color: var(--vscode-foreground);">Write a query that calculates a row number for each order</span>   <span style="color: var(--vscode-foreground);">&nbsp;based on orderdate, orderid ordering</span>

**Tables**: Sales.\[Order\]

**Columns**: orderid, orderdate, customerid, employeeid

**Predicate**: Counts all previous orders up to the current one, based on order date and order ID.

In [50]:
USE TSQLV6
SELECT 
    O1.orderid, 
    O1.orderdate, 
    O1.custid, 
    O1.empid,
    (SELECT COUNT(*) 
     FROM Sales.Orders AS O2
     WHERE O2.orderdate < O1.orderdate OR (O2.orderdate = O1.orderdate AND O2.orderid <= O1.orderid)) AS rownum
FROM 
    Sales.Orders AS O1
ORDER BY 
    O1.orderdate, 
    O1.orderid;

USE Northwinds2022TSQLV7
SELECT 
    O1.orderid, 
    O1.orderdate, 
    O1.customerid, 
    O1.employeeid,
    (SELECT COUNT(*) 
     FROM Sales.[Order] AS O2
     WHERE O2.orderdate < O1.orderdate OR (O2.orderdate = O1.orderdate AND O2.orderid <= O1.orderid)) AS rownum
FROM 
    Sales.[Order] AS O1
ORDER BY 
    O1.orderdate, 
    O1.orderid;


(830 rows affected)

(830 rows affected)

Total execution time: 00:00:00.211

orderid orderdate custid empid rownum 10248 2020-07-04 85 5 1 10249 2020-07-05 79 6 2 10250 2020-07-08 34 4 3 10251 2020-07-08 84 3 4 10252 2020-07-09 76 4 5 10253 2020-07-10 34 3 6 10254 2020-07-11 14 5 7 10255 2020-07-12 68 9 8 10256 2020-07-15 88 3 9 10257 2020-07-16 35 4 10 10258 2020-07-17 20 1 11 10259 2020-07-18 13 4 12 10260 2020-07-19 56 4 13 10261 2020-07-19 61 4 14 10262 2020-07-22 65 8 15 10263 2020-07-23 20 9 16 10264 2020-07-24 24 6 17 10265 2020-07-25 7 2 18 10266 2020-07-26 87 3 19 10267 2020-07-29 25 4 20 10268 2020-07-30 33 8 21 10269 2020-07-31 89 5 22 10270 2020-08-01 87 1 23 10271 2020-08-01 75 6 24 10272 2020-08-02 65 6 25 10273 2020-08-05 63 3 26 10274 2020-08-06 85 6 27 10275 2020-08-07 49 1 28 10276 2020-08-08 80 8 29 10277 2020-08-09 52 2 30 10278 2020-08-12 5 8 31 10279 2020-08-13 44 8 32 10280 2020-08-14 5 2 33 10281 2020-08-14 69 4 34 10282 2020-08-15 69 4 35 10283 2020-08-16 46 3 36 10284 2020-08-19 44 4 37 10285 2020-08-20 63 1 38 10286 2020-08-21 63 8 39 10287 2020-08-22 67 8 40 10288 2020-08-23 66 4 41 10289 2020-08-26 11 7 42 10290 2020-08-27 15 8 43 10291 2020-08-27 61 6 44 10292 2020-08-28 81 1 45 10293 2020-08-29 80 1 46 10294 2020-08-30 65 4 47 10295 2020-09-02 85 2 48 10296 2020-09-03 46 6 49 10297 2020-09-04 7 5 50 10298 2020-09-05 37 6 51 10299 2020-09-06 67 4 52 10300 2020-09-09 49 2 53 10301 2020-09-09 86 8 54 10302 2020-09-10 76 4 55 10303 2020-09-11 30 7 56 10304 2020-09-12 80 1 57 10305 2020-09-13 55 8 58 10306 2020-09-16 69 1 59 10307 2020-09-17 48 2 60 10308 2020-09-18 2 7 61 10309 2020-09-19 37 3 62 10310 2020-09-20 77 8 63 10311 2020-09-20 18 1 64 10312 2020-09-23 86 2 65 10313 2020-09-24 63 2 66 10314 2020-09-25 65 1 67 10315 2020-09-26 38 4 68 10316 2020-09-27 65 1 69 10317 2020-09-30 48 6 70 10318 2020-10-01 38 8 71 10319 2020-10-02 80 7 72 10320 2020-10-03 87 5 73 10321 2020-10-03 38 3 74 10322 2020-10-04 58 7 75 10323 2020-10-07 39 4 76 10324 2020-10-08 71 9 77 10325 2020-10-09 39 1 78 10326 2020-10-10 8 4 79 10327 2020-10-11 24 2 80 10328 2020-10-14 28 4 81 10329 2020-10-15 75 4 82 10330 2020-10-16 46 3 83 10331 2020-10-16 9 9 84 10332 2020-10-17 51 3 85 10333 2020-10-18 87 5 86 10334 2020-10-21 84 8 87 10335 2020-10-22 37 7 88 10336 2020-10-23 60 7 89 10337 2020-10-24 25 4 90 10338 2020-10-25 55 4 91 10339 2020-10-28 51 2 92 10340 2020-10-29 9 1 93 10341 2020-10-29 73 7 94 10342 2020-10-30 25 4 95 10343 2020-10-31 44 4 96 10344 2020-11-01 89 4 97 10345 2020-11-04 63 2 98 10346 2020-11-05 65 3 99 10347 2020-11-06 21 4 100 10348 2020-11-07 86 4 101 10349 2020-11-08 75 7 102 10350 2020-11-11 41 6 103 10351 2020-11-11 20 1 104 10352 2020-11-12 28 3 105 10353 2020-11-13 59 7 106 10354 2020-11-14 58 8 107 10355 2020-11-15 4 6 108 10356 2020-11-18 86 6 109 10357 2020-11-19 46 1 110 10358 2020-11-20 41 5 111 10359 2020-11-21 72 5 112 10360 2020-11-22 7 4 113 10361 2020-11-22 63 1 114 10362 2020-11-25 9 3 115 10363 2020-11-26 17 4 116 10364 2020-11-26 19 1 117 10365 2020-11-27 3 3 118 10366 2020-11-28 29 8 119 10367 2020-11-28 83 7 120 10368 2020-11-29 20 2 121 10369 2020-12-02 75 8 122 10370 2020-12-03 14 6 123 10371 2020-12-03 41 1 124 10372 2020-12-04 62 5 125 10373 2020-12-05 37 4 126 10374 2020-12-05 91 1 127 10375 2020-12-06 36 3 128 10376 2020-12-09 51 1 129 10377 2020-12-09 72 1 130 10378 2020-12-10 24 5 131 10379 2020-12-11 61 2 132 10380 2020-12-12 37 8 133 10381 2020-12-12 46 3 134 10382 2020-12-13 20 4 135 10383 2020-12-16 4 8 136 10384 2020-12-16 5 3 137 10385 2020-12-17 75 1 138 10386 2020-12-18 21 9 139 10387 2020-12-18 70 1 140 10388 2020-12-19 72 2 141 10389 2020-12-20 10 4 142 10390 2020-12-23 20 6 143 10391 2020-12-23 17 3 144 10392 2020-12-24 59 2 145 10393 2020-12-25 71 1 146 10394 2020-12-25 36 1 147 10395 2020-12-26 35 6 148 10396 2020-12-27 25 1 149 10397 2020-12-27 60 5 150 10398 2020-12-30 71 2 151 10399 2020-12-31 83 8 152 10400 2021-01-01 19 1 153 10401 2021-01-01 65 1 154 10402 2021-01-02 20 8 155 10403 2021-01-03 20 4 156 10404 2021-01-03 49 2 157 10

orderid orderdate customerid employeeid rownum 10248 2014-07-04 85 5 1 10249 2014-07-05 79 6 2 10250 2014-07-08 34 4 3 10251 2014-07-08 84 3 4 10252 2014-07-09 76 4 5 10253 2014-07-10 34 3 6 10254 2014-07-11 14 5 7 10255 2014-07-12 68 9 8 10256 2014-07-15 88 3 9 10257 2014-07-16 35 4 10 10258 2014-07-17 20 1 11 10259 2014-07-18 13 4 12 10260 2014-07-19 56 4 13 10261 2014-07-19 61 4 14 10262 2014-07-22 65 8 15 10263 2014-07-23 20 9 16 10264 2014-07-24 24 6 17 10265 2014-07-25 7 2 18 10266 2014-07-26 87 3 19 10267 2014-07-29 25 4 20 10268 2014-07-30 33 8 21 10269 2014-07-31 89 5 22 10270 2014-08-01 87 1 23 10271 2014-08-01 75 6 24 10272 2014-08-02 65 6 25 10273 2014-08-05 63 3 26 10274 2014-08-06 85 6 27 10275 2014-08-07 49 1 28 10276 2014-08-08 80 8 29 10277 2014-08-09 52 2 30 10278 2014-08-12 5 8 31 10279 2014-08-13 44 8 32 10280 2014-08-14 5 2 33 10281 2014-08-14 69 4 34 10282 2014-08-15 69 4 35 10283 2014-08-16 46 3 36 10284 2014-08-19 44 4 37 10285 2014-08-20 63 1 38 10286 2014-08-21 63 8 39 10287 2014-08-22 67 8 40 10288 2014-08-23 66 4 41 10289 2014-08-26 11 7 42 10290 2014-08-27 15 8 43 10291 2014-08-27 61 6 44 10292 2014-08-28 81 1 45 10293 2014-08-29 80 1 46 10294 2014-08-30 65 4 47 10295 2014-09-02 85 2 48 10296 2014-09-03 46 6 49 10297 2014-09-04 7 5 50 10298 2014-09-05 37 6 51 10299 2014-09-06 67 4 52 10300 2014-09-09 49 2 53 10301 2014-09-09 86 8 54 10302 2014-09-10 76 4 55 10303 2014-09-11 30 7 56 10304 2014-09-12 80 1 57 10305 2014-09-13 55 8 58 10306 2014-09-16 69 1 59 10307 2014-09-17 48 2 60 10308 2014-09-18 2 7 61 10309 2014-09-19 37 3 62 10310 2014-09-20 77 8 63 10311 2014-09-20 18 1 64 10312 2014-09-23 86 2 65 10313 2014-09-24 63 2 66 10314 2014-09-25 65 1 67 10315 2014-09-26 38 4 68 10316 2014-09-27 65 1 69 10317 2014-09-30 48 6 70 10318 2014-10-01 38 8 71 10319 2014-10-02 80 7 72 10320 2014-10-03 87 5 73 10321 2014-10-03 38 3 74 10322 2014-10-04 58 7 75 10323 2014-10-07 39 4 76 10324 2014-10-08 71 9 77 10325 2014-10-09 39 1 78 10326 2014-10-10 8 4 79 10327 2014-10-11 24 2 80 10328 2014-10-14 28 4 81 10329 2014-10-15 75 4 82 10330 2014-10-16 46 3 83 10331 2014-10-16 9 9 84 10332 2014-10-17 51 3 85 10333 2014-10-18 87 5 86 10334 2014-10-21 84 8 87 10335 2014-10-22 37 7 88 10336 2014-10-23 60 7 89 10337 2014-10-24 25 4 90 10338 2014-10-25 55 4 91 10339 2014-10-28 51 2 92 10340 2014-10-29 9 1 93 10341 2014-10-29 73 7 94 10342 2014-10-30 25 4 95 10343 2014-10-31 44 4 96 10344 2014-11-01 89 4 97 10345 2014-11-04 63 2 98 10346 2014-11-05 65 3 99 10347 2014-11-06 21 4 100 10348 2014-11-07 86 4 101 10349 2014-11-08 75 7 102 10350 2014-11-11 41 6 103 10351 2014-11-11 20 1 104 10352 2014-11-12 28 3 105 10353 2014-11-13 59 7 106 10354 2014-11-14 58 8 107 10355 2014-11-15 4 6 108 10356 2014-11-18 86 6 109 10357 2014-11-19 46 1 110 10358 2014-11-20 41 5 111 10359 2014-11-21 72 5 112 10360 2014-11-22 7 4 113 10361 2014-11-22 63 1 114 10362 2014-11-25 9 3 115 10363 2014-11-26 17 4 116 10364 2014-11-26 19 1 117 10365 2014-11-27 3 3 118 10366 2014-11-28 29 8 119 10367 2014-11-28 83 7 120 10368 2014-11-29 20 2 121 10369 2014-12-02 75 8 122 10370 2014-12-03 14 6 123 10371 2014-12-03 41 1 124 10372 2014-12-04 62 5 125 10373 2014-12-05 37 4 126 10374 2014-12-05 91 1 127 10375 2014-12-06 36 3 128 10376 2014-12-09 51 1 129 10377 2014-12-09 72 1 130 10378 2014-12-10 24 5 131 10379 2014-12-11 61 2 132 10380 2014-12-12 37 8 133 10381 2014-12-12 46 3 134 10382 2014-12-13 20 4 135 10383 2014-12-16 4 8 136 10384 2014-12-16 5 3 137 10385 2014-12-17 75 1 138 10386 2014-12-18 21 9 139 10387 2014-12-18 70 1 140 10388 2014-12-19 72 2 141 10389 2014-12-20 10 4 142 10390 2014-12-23 20 6 143 10391 2014-12-23 17 3 144 10392 2014-12-24 59 2 145 10393 2014-12-25 71 1 146 10394 2014-12-25 36 1 147 10395 2014-12-26 35 6 148 10396 2014-12-27 25 1 149 10397 2014-12-27 60 5 150 10398 2014-12-30 71 2 151 10399 2014-12-31 83 8 152 10400 2015-01-01 19 1 153 10401 2015-01-01 65 1 154 10402 2015-01-02 20 8 155 10403 2015-01-03 20 4 156 10404 2015-01-03 49

\-- 3-2

\-- Write a query that returns rows with row numbers 11 through 20

\-- based on the row number definition in exercise 3-1

\-- Use a CTE to encapsulate the code from exercise 3-1

\-- Tables involved: Sales.Orders

\-- Desired output:

orderid     orderdate   custid      empid       rownum

\----------- ----------- ----------- ----------- -------

10258       2020-07-17  20          1           11

10259       2020-07-18  13          4           12

10260       2020-07-19  56          4           13

10261       2020-07-19  61          4           14

10262       2020-07-22  65          8           15

10263       2020-07-23  20          9           16

10264       2020-07-24  24          6           17

10265       2020-07-25  7           2           18

10266       2020-07-26  87          3           19

10267       2020-07-29  25          4           20

(10 rows affected)

<span style="color: var(--tw-prose-bold); border: 0px solid rgb(227, 227, 227); box-sizing: border-box; --tw-border-spacing-x: 0; --tw-border-spacing-y: 0; --tw-translate-x: 0; --tw-translate-y: 0; --tw-rotate: 0; --tw-skew-x: 0; --tw-skew-y: 0; --tw-scale-x: 1; --tw-scale-y: 1; --tw-pan-x: ; --tw-pan-y: ; --tw-pinch-zoom: ; --tw-scroll-snap-strictness: proximity; --tw-gradient-from-position: ; --tw-gradient-via-position: ; --tw-gradient-to-position: ; --tw-ordinal: ; --tw-slashed-zero: ; --tw-numeric-figure: ; --tw-numeric-spacing: ; --tw-numeric-fraction: ; --tw-ring-inset: ; --tw-ring-offset-width: 0px; --tw-ring-offset-color: #fff; --tw-ring-color: rgba(69,89,164,.5); --tw-ring-offset-shadow: 0 0 transparent; --tw-ring-shadow: 0 0 transparent; --tw-shadow: 0 0 transparent; --tw-shadow-colored: 0 0 transparent; --tw-blur: ; --tw-brightness: ; --tw-contrast: ; --tw-grayscale: ; --tw-hue-rotate: ; --tw-invert: ; --tw-saturate: ; --tw-sepia: ; --tw-drop-shadow: ; --tw-backdrop-blur: ; --tw-backdrop-brightness: ; --tw-backdrop-contrast: ; --tw-backdrop-grayscale: ; --tw-backdrop-hue-rotate: ; --tw-backdrop-invert: ; --tw-backdrop-opacity: ; --tw-backdrop-saturate: ; --tw-backdrop-sepia: ;"><b>Proposition</b></span><span style="color: var(--tw-prose-bold); border: 0px solid rgb(227, 227, 227); box-sizing: border-box; --tw-border-spacing-x: 0; --tw-border-spacing-y: 0; --tw-translate-x: 0; --tw-translate-y: 0; --tw-rotate: 0; --tw-skew-x: 0; --tw-skew-y: 0; --tw-scale-x: 1; --tw-scale-y: 1; --tw-pan-x: ; --tw-pan-y: ; --tw-pinch-zoom: ; --tw-scroll-snap-strictness: proximity; --tw-gradient-from-position: ; --tw-gradient-via-position: ; --tw-gradient-to-position: ; --tw-ordinal: ; --tw-slashed-zero: ; --tw-numeric-figure: ; --tw-numeric-spacing: ; --tw-numeric-fraction: ; --tw-ring-inset: ; --tw-ring-offset-width: 0px; --tw-ring-offset-color: #fff; --tw-ring-color: rgba(69,89,164,.5); --tw-ring-offset-shadow: 0 0 transparent; --tw-ring-shadow: 0 0 transparent; --tw-shadow: 0 0 transparent; --tw-shadow-colored: 0 0 transparent; --tw-blur: ; --tw-brightness: ; --tw-contrast: ; --tw-grayscale: ; --tw-hue-rotate: ; --tw-invert: ; --tw-saturate: ; --tw-sepia: ; --tw-drop-shadow: ; --tw-backdrop-blur: ; --tw-backdrop-brightness: ; --tw-backdrop-contrast: ; --tw-backdrop-grayscale: ; --tw-backdrop-hue-rotate: ; --tw-backdrop-invert: ; --tw-backdrop-opacity: ; --tw-backdrop-saturate: ; --tw-backdrop-sepia: ; font-weight: 600;">:&nbsp;</span> <span style="color: var(--vscode-foreground);">Write a query that returns rows with row numbers 11 through 20&nbsp;</span>  <span style="color: var(--vscode-foreground);">based on the row number definition in exercise 3-1</span>   

**Tables**: Sales.\[Order\]

**Columns**: orderid, orderdate, customerid, employeeid

**Predicate**: Selects orders ranked 11th to 20th by their calculated sequence number, determined by order date and order ID.

In [51]:
USE TSQLV6
;WITH RowNumberedOrders AS 
(
    SELECT 
        O1.orderid, 
        O1.orderdate, 
        O1.custid, 
        O1.empid,
        (SELECT COUNT(*) 
         FROM Sales.Orders AS O2
         WHERE O2.orderdate < O1.orderdate OR (O2.orderdate = O1.orderdate AND O2.orderid <= O1.orderid)) AS rownum
    FROM 
        Sales.Orders AS O1
)
SELECT 
    orderid, 
    orderdate, 
    custid, 
    empid,
    rownum
FROM 
    RowNumberedOrders
WHERE 
    rownum BETWEEN 11 AND 20
ORDER BY 
    rownum;

USE Northwinds2022TSQLV7
;WITH RowNumberedOrders AS 
(
    SELECT 
        O1.orderid, 
        O1.orderdate, 
        O1.customerid, 
        O1.employeeid,
        (SELECT COUNT(*) 
         FROM Sales.[Order] AS O2
         WHERE O2.orderdate < O1.orderdate OR (O2.orderdate = O1.orderdate AND O2.orderid <= O1.orderid)) AS rownum
    FROM 
        Sales.[Order] AS O1
)
SELECT 
    orderid, 
    orderdate, 
    customerid, 
    employeeid,
    rownum
FROM 
    RowNumberedOrders
WHERE 
    rownum BETWEEN 11 AND 20
ORDER BY 
    rownum;

(10 rows affected)

(10 rows affected)

Total execution time: 00:00:00.165

orderid orderdate custid empid rownum 10258 2020-07-17 20 1 11 10259 2020-07-18 13 4 12 10260 2020-07-19 56 4 13 10261 2020-07-19 61 4 14 10262 2020-07-22 65 8 15 10263 2020-07-23 20 9 16 10264 2020-07-24 24 6 17 10265 2020-07-25 7 2 18 10266 2020-07-26 87 3 19 10267 2020-07-29 25 4 20

orderid orderdate customerid employeeid rownum 10258 2014-07-17 20 1 11 10259 2014-07-18 13 4 12 10260 2014-07-19 56 4 13 10261 2014-07-19 61 4 14 10262 2014-07-22 65 8 15 10263 2014-07-23 20 9 16 10264 2014-07-24 24 6 17 10265 2014-07-25 7 2 18 10266 2014-07-26 87 3 19 10267 2014-07-29 25 4 20

**NACE COMPENTENCIES and GPT:**

When working as a team, I naturally used and developed my skills in **leadership**, **teamwork**, and **communication**. I continued to collaborate and communicate efficiently with my team, we developed a clear plan in tackling this assignment, and kept my team up to date with my progress.

When working individually, I developed and used my skills in **critical thinking** and **technology**. I developed my SQL skills by reading Chapter 4 and 5. I had to analyze the problems that were given, break them down, and create effective solutions, I then adapted these solutions into the Northwinds database.

When creating the MP4 file I also had to utilize **technology,** I've already gained a good sense on how to use OBS for video recording. Additionally I had to collaborate with my team and make sure we had a clear plan in developing this video, which utilized **teamwork** and **communication**.

As for ChatGPT, I used ChatGPT again to help me quickly translate the TSQLV6 queries to work with the Northern database, GPT made replacing table and column names very fast.

_Written in collaboration with ChatGPT from OpenAI to improve understanding and assist with the explanation of the query._